File: java/dp/unique-paths.java
class Solution {
    int [][] dp;
    int row; 
    int col;

    public int uniquePaths(int m, int n) {
        dp = new int[m][n];
        row = m;
        col = n;
        for(int i=0;i<m;i++){
            Arrays.fill(dp[i],-1);
        }
        return path(0,0);
    }
    public int path(int i, int j){
        if(i>=row || j >= col ){
            return 0;
        }
        if (i == row-1 && j == col-1){
            return 1;
        }
        if(dp[i][j] != -1){
            return dp[i][j];
        }

        dp[i][j] = path(i+1,j) + path(i,j+1);
        return dp[i][j];
    }
}

File: python/2d_dynamic_programming/unique_paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        cache = [[-1 for _ in range(n)] for _ in range(m)]

        def dfs(i, j):
            if i == m or j == n:
                return 0
            if i == m-1 or j == n-1:
                return 1
            if cache[i][j] != -1:
                return cache[i][j]
            
            cache[i][j] = dfs(i+1, j) + dfs(i, j+1)

            return cache[i][j]

        return dfs(0, 0)